lib/r10k/action/deploy/environment.rb

```diff
@@ -43,6 +45,38 @@ def call
 
         private
 
+        def extract_credentials!
+          if @sshkey_path && @token_path
+            raise R10K::Error, "Cannot specify both an SSH key and a token to use with this deploy."
+          end
+
+          if @sshkey_path
+            if File.exists?(@sshkey_path)
+              { sshkey_path: @sshkey_path }
+            else
+              raise R10K::Error, _("{%path} does not exist, cannot load SSH key") % { path: @sshkey_path }
+            end
+          elsif @token_path
+            if @token_path == '-'
+              token = $stdin.read
+            elsif File.exists?(@token_path)
+              token = File.read(@token_path).strip
+            else
+              raise R10K::Error, _("{%path} does not exist, cannot load OAuth token") % { path: @token_path }
+            end
+
+            unless valid_token?(token)
+              raise R10K::Error, _("Supplied token contains invalid characters.")
+            end
+
+            { token: token }
+          end
+        end
+
+        def valid_token?(token)
+          return token =~ /^[\w\-\.~\+\/]+$/
```

Comment:
It would be valuable to add a comment pointing to your link (https://www.oauth.com/oauth2-servers/access-tokens/access-token-response/) about why we're using this bizarre set of characters.